openembedded-core-mbl/scripts/lib/mbl-signing/signing/pki/backends.py

```diff
@@ -0,0 +1,257 @@
+# Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+
+"""
+This module defines a the "backend" interface for secrets storage engines.
+
+The interface allows the storage engine to be called by the signing REST API.
+
+If you want to configure your own secrets backend, you must implement the
+SigningBackend interface.
+
+However this is not fully supported in this version of the API.
+But it will be in the release version.
+
+The implemented backend is a wrapper around a Hashicorp Vault client.
+For more information on the Vault PKI engine see the following link:
+
+https://www.vaultproject.io/api/secret/pki/index.html
+"""
+
+from abc import abstractmethod, ABC
+import hvac
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography import x509
+from requests import ConnectionError
+
+
+def signing_backend():
+    """Return the chosen signing backend."""
+    return VaultBackend()
+
+
+class SigningBackend(ABC):
+    """Abstract interface for signing backends.
+
+    To create a backend all of the abstract methods must be overriden.
+    You can define your own method signatures, however these must be
+    compatible with the REST API definition.
+
+    An example implementation is in the signing.pki.endpoints module
+    """
+
+    @property
+    @abstractmethod
+    def server_url(self):
+        """URL of the signing backend server."""
+
+    @abstractmethod
+    def generate_root(self, *args, **kwargs):
+        """Generate a root CA."""
+
+    @abstractmethod
+    def generate_intermediate(self, *args, **kwargs):
+        """Generate an intermediate CA."""
+
+    @abstractmethod
+    def create_or_update_role(self, *args, **kwargs):
+        """
+        Create or update a role.
+
+        Roles are a set of parameters used to create the certificate.
+        These parameters include TTL, allowed common names, key usage etc.
+        """
+
+    @abstractmethod
+    def request_certificate(self, *args, **kwargs):
+        """Request an end entity certificate."""
+
+    @abstractmethod
+    def read_certificate(self, *args, **kwargs):
+        """Read a stored certificate from the backend."""
+
+    @abstractmethod
+    def list_certificates(self, *args, **kwargs):
+        """List certificates stored in the backend."""
+
+
+class VaultBackend(SigningBackend):
+    """
+    Implementation of the SigningBackend interface.
+
+    The backend used in this case is Hashicorp Vault.
```

Comment:
I think it might be worth explaining the mount_point system (as I presume this is Vault specific).
Though - this is probably what you have written up in your document explaining the flow, so maybe its not needed here.

Reply:
I'm attempting to abstract away the 'mount' system in my API. The "front end" of the API doesn't mention mount points anywhere (or shouldn't!). I'm only using that terminology in the "backend" example. I will explain how Vault works in the docs though.